pkg/maglev/maglev.go

```diff
@@ -56,21 +60,60 @@ type Config struct {
 	MaglevHashSeed string `mapstructure:"bpf-lb-maglev-hash-seed"`
 }
 
-func (def Config) Flags(flags *pflag.FlagSet) {
+func (def UserConfig) Flags(flags *pflag.FlagSet) {
 	flags.Uint(MaglevTableSizeName, def.MaglevTableSize, fmt.Sprintf("Maglev per service backend table size (parameter M, one of: %v)", maglevSupportedTableSizes))
 	flags.String(MaglevHashSeedName, def.MaglevHashSeed, "Maglev cluster-wide hash seed (base64 encoded)")
 }
 
-var DefaultConfig = Config{
+func (userCfg UserConfig) ToConfig() (Config, error) {
+	if !slices.Contains(maglevSupportedTableSizes, userCfg.MaglevTableSize) {
+		return Config{}, fmt.Errorf("Invalid value for --%s: %d, supported values are: %v",
+			MaglevTableSizeName, userCfg.MaglevTableSize, maglevSupportedTableSizes)
+	}
+
+	seed := userCfg.MaglevHashSeed
+	d, err := base64.StdEncoding.DecodeString(seed)
+	if err != nil {
+		return Config{}, fmt.Errorf("cannot decode base64 Maglev hash seed %q: %w", seed, err)
+	}
+	if len(d) != 12 {
+		return Config{}, fmt.Errorf("decoded hash seed is %d bytes (not 12 bytes)", len(d))
+	}
+	return Config{
+		MaglevTableSize: userCfg.MaglevTableSize,
+		MaglevHashSeed:  userCfg.MaglevHashSeed,
+		SeedMurmur:      uint32(d[0])<<24 | uint32(d[1])<<16 | uint32(d[2])<<8 | uint32(d[3]),
+		SeedJhash0:      uint32(d[4])<<24 | uint32(d[5])<<16 | uint32(d[6])<<8 | uint32(d[7]),
+		SeedJhash1:      uint32(d[8])<<24 | uint32(d[9])<<16 | uint32(d[10])<<8 | uint32(d[11]),
+	}, nil
+}
+
+var DefaultUserConfig = UserConfig{
 	MaglevTableSize: DefaultTableSize,
 	MaglevHashSeed:  DefaultHashSeed,
 }
 
-type Maglev struct {
-	Config
+// DefaultConfig is the default maglev configuration for testing.
+var DefaultConfig, _ = DefaultUserConfig.ToConfig()
+
+// Config is the maglev configuration derived from the user configuration.
+type Config struct {
+	// Maglev backend table size (M) per service. Must be prime number.
+	// "Let N be the size of a VIP's backend pool." [...] "In practice, we choose M to be
+	// larger than 100 x N to ensure at most a 1% difference in hash space assigned to
+	// backends." (from Maglev paper, page 6)
+	MaglevTableSize uint
```

Comment:
Nit: I'd call this `TableSize`, since it stutters at the call site (e.g. `cfg.MaglevConfig.MaglevTableSize`). Same for `MaglevHashSeed` and the equivalent fields in UserConfig. Not sure how much churn this would cause in the codebase, though.

Reply:
Oh yeah good point. Dropping the prefix seems nicer to me as well so I'll go ahead and clean that up.